calicoctl/commands/migrate/import.go

```diff
@@ -0,0 +1,409 @@
+	}
+
+	// TODO: On failure, print instructions on how to cleanse datastore
+	err = checkCalicoResourcesNotExist(parsedArgs, client)
+	if err != nil {
+		// TODO: Add something like 'calicoctl migrate clean' to delete all the CRDs to wipe out the Calico resources.
+		return fmt.Errorf("Datastore already has Calico resources: %s. Clear out all Calico resources by deleting all Calico CRDs.", err)
+	}
+
+	// Ensure that the cluster info resource is initialized.
+	ctx := context.Background()
+	if err := client.EnsureInitialized(ctx, "", ""); err != nil {
+		return fmt.Errorf("Unable to initialize cluster information for the datastore migration: %s", err)
+	}
+
+	// Make sure that the datastore is locked. Since the call to EnsureInitialized
+	// should initialize it to unlocked, lock it before we continue.
+	locked, err := checkLocked(ctx, client)
+	if err != nil {
+		return fmt.Errorf("Error while checking if datastore was locked: %s", err)
+	} else if !locked {
+		err := Lock([]string{"migrate", "lock", "-c", cf})
+		if err != nil {
+			return fmt.Errorf("Error while attempting to lock the datastore for import: %s", err)
+		}
+	}
+
+	// Split file into v3 API, ClusterGUID, and IPAM components
+	filename := parsedArgs["--filename"].(string)
+	v3Yaml, clusterInfoJson, ipamJson, err := splitImportFile(filename)
+	if err != nil {
+		return fmt.Errorf("Error while reading migration file: %s\n", err)
+	}
+
+	// Apply v3 API resources
+	err = updateV3Resources(cf, v3Yaml)
+	if err != nil {
+		return fmt.Errorf("Failed to import v3 resources: %s\n", err)
+	}
+
+	// Update the clusterinfo resource with the data from the old datastore.
+	err = updateClusterInfo(ctx, client, clusterInfoJson)
+	if err != nil {
+		return fmt.Errorf("Failed to update cluster information: %s", err)
+	}
+
+	// Import IPAM components
+	fmt.Print("Importing IPAM resources\n")
+	ipam := NewMigrateIPAM(client)
+	err = json.Unmarshal(ipamJson, ipam)
+	if err != nil {
+		return fmt.Errorf("Failed to read IPAM resources: %s\n", err)
+	}
+	results := ipam.PushToDatastore()
+
+	// Handle the IPAM results
+	if results.numHandled == 0 {
+		if results.numResources == 0 {
+			return fmt.Errorf("No IPAM resources specified in file")
+		} else {
+			return fmt.Errorf("Failed to import any IPAM resources: %v", results.resErrs)
+		}
+	} else if len(results.resErrs) == 0 {
+		fmt.Printf("Successfully applied %d IPAM resource(s)\n", results.numHandled)
+	} else {
+		if results.numHandled != 0 && len(results.resErrs) > 0 {
+			fmt.Printf("Partial success: ")
+			fmt.Printf("applied the first %d out of %d resources:\n", results.numHandled, results.numResources)
+		}
+		return fmt.Errorf("Hit error(s): %v", results.resErrs)
+	}
+
+	// TODO: Be more specific on how to do the calico configuration change.
+	fmt.Print("Datastore information successfully imported. To complete the datastore migration, run `calicoctl unlock` and modify your calico configuration to match the kubernetes datastore.\n")
```

Comment:
I don't think we should print this out here - unlocking the data store should only happen after the rolling update is complete.

Reply:
Would swapping the order of the instructions work? I feel like it is useful to be as prescriptive as possible since people may forget to unlock the datastore.

Reply:
But barring that, I think it should just say "refer to the upgrade docs for next steps"